packages/opencode/src/session/prompt.ts

```diff
@@ -195,6 +196,66 @@ export namespace SessionPrompt {
       processor,
     })
 
+    const permUnsub = (() => {
+      if (!input.acpConnection) return
+      const acp = input.acpConnection
+      const handled = new Set<string>()
+      const options = [
+        { optionId: "allow_once", kind: "allow_once", name: "Allow once" },
+        { optionId: "allow_always", kind: "allow_always", name: "Always allow" },
+        { optionId: "reject_once", kind: "reject_once", name: "Reject" },
+      ]
+      return Bus.subscribe(Permission.Event.Updated, (event) => {
+        const info = event.properties
+        if (info.sessionID !== input.sessionID) return
+        if (handled.has(info.id)) return
+        handled.add(info.id)
+        const toolCallId = info.callID ?? info.id
+        const metadata = info.metadata ?? {}
+        return acp.connection
+          .requestPermission({
+            sessionId: acp.sessionId,
+            toolCall: {
+              toolCallId,
+              status: "pending",
+              title: info.title,
+              rawInput: metadata,
+              kind: permissionTypeToToolKind(info.type),
+              locations: extractLocations(info.type, metadata),
+            },
+            options,
+          })
+          .then((response: RequestPermissionResponse) => {
+            if (response.outcome.outcome !== "selected") {
+              Permission.respond({ sessionID: info.sessionID, permissionID: info.id, response: "reject" })
+              return
+            }
+            const optionId = response.outcome.optionId
+            if (optionId === "allow_once") {
+              Permission.respond({ sessionID: info.sessionID, permissionID: info.id, response: "once" })
+              return
+            }
+            if (optionId === "allow_always") {
+              Permission.respond({ sessionID: info.sessionID, permissionID: info.id, response: "always" })
+              return
+            }
+            Permission.respond({ sessionID: info.sessionID, permissionID: info.id, response: "reject" })
+          })
+          .catch((error: Error) => {
+            log.error("failed to request permission from ACP", {
+              error,
+              permissionID: info.id,
+              sessionID: info.sessionID,
+            })
+            Permission.respond({ sessionID: info.sessionID, permissionID: info.id, response: "reject" })
+          })
+      })
+    })()
+    await using _permSub = defer(() => {
+      if (!permUnsub) return
+      permUnsub()
+    })
```

Comment:
Consider using optional chain: 
```suggestion
    await using _permSub = defer(() => {
      permUnsub?.()
    })
```

Reply:
Good suggestion, it has been modified.